---
title: Comparing DNAm sites close to candidate eczema genes to those in other genomic regions
author: Thomas Battram
output:
  bookdown::html_document2:
    toc: true
    toc_float: true
    theme: united
    number_sections: false
params:
  filenames: ""
  plots: ""
---

```{r setup, message=FALSE, warning=FALSE, include=FALSE}

## pkgs
suppressWarnings(suppressPackageStartupMessages({
library(bookdown) # extra markdown options
library(knitr) # ...
library(kableExtra) # extra nice tables
library(tidyverse) # tidy code and data
library(cowplot) # plotting figures on a grid
library(RColorBrewer) # for plotting colours
library(usefunc) # own package of useful functions
}))

knitr::opts_chunk$set(echo = FALSE, 
           			  warning = FALSE, 
            		  message = TRUE, 
            		  cache = FALSE, 
            		  dpi = 300 
            		  )

# Function for numbers in rmarkdown
# comma <- function(x) format(x, digits = 2, big.mark = ",")
```

```{r general-setup}
# params <- list(filenames = "results/m1c-comp-summary.RData results/m2c-comp-summary.RData results/m3c-comp-summary.RData", plots = "results/plots/m1c-qq-comparison.png results/plots/m2c-qq-comparison.png results/plots/m3c-qq-comparison.png")
get_model <- function(res_file)
{
    stringr::str_extract(res_file, "m[1-3][a-c]")
}

files <- unlist(str_split(params$filenames, " "))
plots <- unlist(str_split(params$plots, " "))
names(plots) <- sapply(plots, get_model)

cpg_total <- new_load(files[1])[["cpg_n"]]

tidy_nums <- function(df) 
{
    df[] <- lapply(df, comma)
    return(df)
}
```

## Summary of project + methods

This study is examining the association between DNA methylation and childhood eczema (up to ~10 years). Studies have data from different timepoints, so in the first instance we will classify individuals as ‘any eczema’ versus ‘no eczema’ according to all timepoints available up to age 10. Sub-analyses will be conducted that limits cases to those children who are diagnosed early in childhood (by ~2 years) or have persistent eczema (diagnosed by age 2 and have current eczema at age 8yrs).

The analyses will therefore use 3 binary definitions:

* [Childhood AD] - cases: AD by age 10, controls:no AD by age 10
* [Early-onset AD] - cases: AD by age 2, controls: no AD by age 2
* [Persistent AD] - cases: AD by age 2 and persisting to ~ 8-10yrs, controls: no AD by age 10

The association of each outcome with DNA methylation will be explored using logistic regression and 3 models with different covariates:

* model a: sex + batch (+ selection_factors) (+ ancestry)
* model b: sex + batch + maternal_age + maternal_smoking + maternal_SES + gestational_age (+selection_factors) (+ ancestry)
* model c: sex + batch + maternal_age + maternal_smoking + maternal_SES + gestational_age + cell_type (+ selection_factors) (+ ancestry)

Results of the meta-analysis are in another report (meta-analysis-report.html). 

This report contains a comparison between DNA methylation sites within 1Mb of genes thought to be involved in eczema development. These genes were identified in a previous paper by [Sobczyk M et al.](https://www.sciencedirect.com/science/article/pii/S0022202X2101160X). Sobczyk et al. use various metrics to assign "prioritization scores" to genes based on GWAS results. We took the 26 top ranked genes and extracted the `r cpg_total` CpGs within 1Mb of those genes. To examine if changes in DNAm at these CpGs tend to be associated more strongly with eczema than other CpGs we made QQ plots from the EWAS meta-analysis results at these CpGs and at all other CpGs. Results can be seen below for model C across each phenotype.

## Childhood AD

```{r qq-plots-m1, results = "asis"}
include_graphics(plots["m1c"])
```

## Early-onset AD

```{r qq-plots-m2, results = "asis"}
include_graphics(plots["m2c"])
```

## Persistent AD

```{r qq-plots-m3, results = "asis"}
include_graphics(plots["m3c"])
```
